# 数据科学家2023年薪资数据

简介：数据集包含数据科学领域不同岗位的薪资水平。
变量含义：
- work_year：支付薪资的年份
- experience_level：工作年份期间的经验水平
- employment_type：职位的就业类型
- job_title：工作年份期间从事的职位
- salary：支付的总薪资金额
- salary_currency：支付的薪资货币，使用 ISO 4217 货币代码表示
- salaryinusd：以美元计算的薪资金额
- employee_residence：员工在工作年份期间的主要居住国家，使用 ISO 3166 国家代码表示
- remote_ratio：远程工作的整体比例
- company_location：雇主总部所在国家或分支机构所在国家
- company_size：公司在工作年份期间的员工人数的中位数

## 读取数据

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv(r"C:\Users\35731\数据科学家2023年薪资数据清洗\ds_salaries.csv")

In [3]:
original_data.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

## 评估整洁度

In [4]:
original_data.sample(10)

work_year experience_level employment_type                   job_title  \
2620       2022               SE              FT     Data Science Consultant   
982        2023               SE              FT                Data Analyst   
401        2023               SE              FT          Analytics Engineer   
474        2023               SE              FT                Data Analyst   
2307       2022               SE              FT               Data Engineer   
3418       2022               MI              FT  Machine Learning Developer   
1824       2022               SE              FT               Data Engineer   
3601       2021               SE              FT          Big Data Architect   
2296       2022               SE              FT                Data Analyst   
2691       2022               MI              FT          Analytics Engineer   

      salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
2620  139000             USD         139000                 US             0   
982   112000             USD         112000                 US           100   
401   179820             USD         179820                 US             0   
474   100000             USD         100000                 US           100   
2307  100000             USD         100000                 US             0   
3418  100000             CAD          76814                 CA           100   
1824  153600             USD         153600                 US             0   
3601  125000             CAD          99703                 CA            50   
2296  177000             USD         177000                 US             0   
2691   85000             USD          85000                 US           100   

     company_location company_size  
2620               US            M  
982                US            M  
401                US            M  
474                US            M  
2307               US            M  
3418               CA            M  
1824               US            M  
3601               CA            M  
2296               US            M  
2691               US            M

其符合“每列一个变量，每行一个观察值，每个单元格一个数据”，具体为每行是一个岗位的信息，每列是关于科学家的一个变量，因此不存在结构性问题

## 评估干净度

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


3755行，无数据缺失，work_year、experience_level、employment_type、company_size的数据类型有误

### 评估缺失数据

无缺失数据

### 评估重复数据

都可以重复

### 评估不一致数据

In [6]:
original_data["company_location"].value_counts()

company_location
US    3040
GB     172
CA      87
ES      77
IN      58
      ... 
CN       1
NZ       1
CL       1
MD       1
MT       1
Name: count, Length: 72, dtype: int64

无不一致

### 评估无效或错误数据

In [7]:
original_data.describe()

work_year        salary  salary_in_usd  remote_ratio
count  3755.000000  3.755000e+03    3755.000000   3755.000000
mean   2022.373635  1.906956e+05  137570.389880     46.271638
std       0.691448  6.716765e+05   63055.625278     48.589050
min    2020.000000  6.000000e+03    5132.000000      0.000000
25%    2022.000000  1.000000e+05   95000.000000      0.000000
50%    2022.000000  1.380000e+05  135000.000000      0.000000
75%    2023.000000  1.800000e+05  175000.000000    100.000000
max    2023.000000  3.040000e+07  450000.000000    100.000000

数据无错误

## 清理数据

- work_year
- experience_level
- employment_type
- company_size
的数据类型有误

In [8]:
cleaned_data = original_data.copy()

In [9]:
cleaned_data.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [10]:
cleaned_data["work_year"] = pd.to_datetime(cleaned_data["work_year"])
cleaned_data["work_year"]

0      1970-01-01 00:00:00.000002023
1      1970-01-01 00:00:00.000002023
2      1970-01-01 00:00:00.000002023
3      1970-01-01 00:00:00.000002023
4      1970-01-01 00:00:00.000002023
                    ...             
3750   1970-01-01 00:00:00.000002020
3751   1970-01-01 00:00:00.000002021
3752   1970-01-01 00:00:00.000002020
3753   1970-01-01 00:00:00.000002020
3754   1970-01-01 00:00:00.000002021
Name: work_year, Length: 3755, dtype: datetime64[ns]

In [17]:
cleaned_data["experience_level"] = cleaned_data["experience_level"].astype("category")

In [16]:
cleaned_data["employment_type"] = cleaned_data["employment_type"].astype("category")

In [15]:
cleaned_data["company_size"] = cleaned_data["company_size"].astype("category")

In [18]:
cleaned_data.describe()

work_year        salary  salary_in_usd  \
count                           3755  3.755000e+03    3755.000000   
mean   1970-01-01 00:00:00.000002022  1.906956e+05  137570.389880   
min    1970-01-01 00:00:00.000002020  6.000000e+03    5132.000000   
25%    1970-01-01 00:00:00.000002022  1.000000e+05   95000.000000   
50%    1970-01-01 00:00:00.000002022  1.380000e+05  135000.000000   
75%    1970-01-01 00:00:00.000002023  1.800000e+05  175000.000000   
max    1970-01-01 00:00:00.000002023  3.040000e+07  450000.000000   
std                              NaN  6.716765e+05   63055.625278   

       remote_ratio  
count   3755.000000  
mean      46.271638  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%      100.000000  
max      100.000000  
std       48.589050

In [19]:
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   work_year           3755 non-null   datetime64[ns]
 1   experience_level    3755 non-null   category      
 2   employment_type     3755 non-null   category      
 3   job_title           3755 non-null   object        
 4   salary              3755 non-null   int64         
 5   salary_currency     3755 non-null   object        
 6   salary_in_usd       3755 non-null   int64         
 7   employee_residence  3755 non-null   object        
 8   remote_ratio        3755 non-null   int64         
 9   company_location    3755 non-null   object        
 10  company_size        3755 non-null   category      
dtypes: category(3), datetime64[ns](1), int64(3), object(4)
memory usage: 246.3+ KB


## 保持数据

In [20]:
cleaned_data.head()

work_year experience_level employment_type  \
0 1970-01-01 00:00:00.000002023               SE              FT   
1 1970-01-01 00:00:00.000002023               MI              CT   
2 1970-01-01 00:00:00.000002023               MI              CT   
3 1970-01-01 00:00:00.000002023               SE              FT   
4 1970-01-01 00:00:00.000002023               SE              FT   

                  job_title  salary salary_currency  salary_in_usd  \
0  Principal Data Scientist   80000             EUR          85847   
1               ML Engineer   30000             USD          30000   
2               ML Engineer   25500             USD          25500   
3            Data Scientist  175000             USD         175000   
4            Data Scientist  120000             USD         120000   

  employee_residence  remote_ratio company_location company_size  
0                 ES           100               ES            L  
1                 US           100               US            S  
2                 US           100               US            S  
3                 CA           100               CA            M  
4                 CA           100               CA            M

In [21]:
cleaned_data.to_csv(r"数据科学家薪资数据已清洗版")

In [22]:
pd.read_csv(r"数据科学家薪资数据已清洗版")

Unnamed: 0                      work_year experience_level  \
0              0  1970-01-01 00:00:00.000002023               SE   
1              1  1970-01-01 00:00:00.000002023               MI   
2              2  1970-01-01 00:00:00.000002023               MI   
3              3  1970-01-01 00:00:00.000002023               SE   
4              4  1970-01-01 00:00:00.000002023               SE   
...          ...                            ...              ...   
3750        3750  1970-01-01 00:00:00.000002020               SE   
3751        3751  1970-01-01 00:00:00.000002021               MI   
3752        3752  1970-01-01 00:00:00.000002020               EN   
3753        3753  1970-01-01 00:00:00.000002020               EN   
3754        3754  1970-01-01 00:00:00.000002021               SE   

     employment_type                 job_title   salary salary_currency  \
0                 FT  Principal Data Scientist    80000             EUR   
1                 CT               ML Engineer    30000             USD   
2                 CT               ML Engineer    25500             USD   
3                 FT            Data Scientist   175000             USD   
4                 FT            Data Scientist   120000             USD   
...              ...                       ...      ...             ...   
3750              FT            Data Scientist   412000             USD   
3751              FT  Principal Data Scientist   151000             USD   
3752              FT            Data Scientist   105000             USD   
3753              CT     Business Data Analyst   100000             USD   
3754              FT      Data Science Manager  7000000             INR   

      salary_in_usd employee_residence  remote_ratio company_location  \
0             85847                 ES           100               ES   
1             30000                 US           100               US   
2             25500                 US           100               US   
3            175000                 CA           100               CA   
4            120000                 CA           100               CA   
...             ...                ...           ...              ...   
3750         412000                 US           100               US   
3751         151000                 US           100               US   
3752         105000                 US           100               US   
3753         100000                 US           100               US   
3754          94665                 IN            50               IN   

     company_size  
0               L  
1               S  
2               S  
3               M  
4               M  
...           ...  
3750            L  
3751            L  
3752            S  
3753            L  
3754            L  

[3755 rows x 12 columns]